rasa_nlu/extractors/nested_entity_extractor.py

```diff
@@ -0,0 +1,264 @@
+             model_dir=None,  # type: Optional[Text]
+             model_metadata=None,  # type: Optional[Metadata]
+             cached_component=None,  # type: Optional[NestedEntitiesMapper]
+             **kwargs  # type: **Any
+             ):
+            # type: (...) -> NestedEntitiesMapper
+
+        meta = model_metadata.for_component(cls.name)
+        file_name = meta.get("nested_entities_file", NESTED_ENTITIES_FILE_NAME)
+        nested_entities_file = os.path.join(model_dir, file_name)
+
+        if os.path.isfile(nested_entities_file):
+            nested_entities = utils.read_json_file(nested_entities_file)
+        else:
+            nested_entities = {
+                'lookup_tables': [],
+                'composite_entities': []
+            }
+            warnings.warn("Failed to load nested entities file from '{}'"
+                          "".format(nested_entities_file))
+
+        return cls(meta, nested_entities)
+
+    def get_relevance(self,
+                      broad_value,
+                      nested_composite_examples):
+        relevance_score = 0
+        for example in nested_composite_examples:
+            if example in broad_value:
+                relevance_score += 1
+        return relevance_score
+
+    def merge_two_dicts(self, x, y):
+        z = x.copy()
+        z.update(y)
+        return z
+
+    def break_on_lookup_tables(self, each_lookup,
+                               child_name,
+                               broad_value):
+        for lookup_entry in each_lookup['elements']:
+            find_index = broad_value.find(lookup_entry.lower())
+            if(find_index > -1):
+                return {
+                    child_name: lookup_entry
+                }
+        return {}
+
+    def split_by_lookup_tables(self, composite_child, broad_value):
+        broken_entity = {}
+        child_name = composite_child[1:]
+        for each_lookup in self.nested_entities['lookup_tables']:
+            if(each_lookup['name'] == child_name):
+                broken_entity = self.merge_two_dicts(
+                    broken_entity,
+                    self.break_on_lookup_tables(each_lookup,
+                                                child_name,
+                                                broad_value))
+        return broken_entity
+
+    def find_number_in_words(self, word):
+        match = False
+        try:
+            match = w2n.word_to_num(builtin_str(word))
+        except ValueError:
+            pass
+        return match
+
+    def find_number_by_regex(self, child_name, broad_value):
+        match = False
+        expression = r'\d+'
+        if(child_name == 'year'):
+            expression = r'\d{4}'
+        match = re.findall(expression, broad_value)
+        if match:
+            match = match[0]
+        return match
+
+    def split_by_sys(self, composite_child, broad_value):
+        broken_entity = {}
+        if(composite_child in ['@number', '@year']):
```

Comment:
This seems a little specific, what is the reason for this function?

Reply:
The name of the function is a bit off. Changed it to `split_number`
The function is meant to extract numbers out from the broad entity if you added it to the list of things to be extracted
I added `year` since some users might want to extract years out and a year is a number